File: Python/Searching/LinearSearch/MoveToFrontHeadLinearSearch.py
# Python3 program for transposition to
# improve the Linear Search Algorithm


# Structure of the array
class Array :
	def __init__(self,a=[0]*10,size=10,l=0) -> None:
		self.A=a
		self.size=size
		self.length=l

# Function to print array element
def Display(arr):

	# Traverse the array arr[]
	for i in range(arr.length) :
		print(arr.A[i],end=" ")
	
	print()


# Function that performs the move to
# front operation for Linear Search
def LinearSearchMoveToFront(arr, key:int):

	# Traverse the array
	for i in range(arr.length) :

		# If key is found, then swap
		# the element with 0-th index
		if (key == arr.A[i]) :
			arr.A[i], arr.A[0]=arr.A[0],arr.A[i]
			return i
		
	
	return -1


# Driver Code
if __name__ == '__main__':
	# Given array arr[]
	arr=Array([2, 23, 14, 5, 6, 9, 8, 12], 10, 8)

	print("Elements before Linear Search Transposition: ",end='')

	# Function Call for displaying
	# the array arr[]
	Display(arr)

	# Function Call for transposition
	LinearSearchMoveToFront(arr, 14)

	print("Elements after Linear Search Transposition: ",end='')

	# Function Call for displaying
	# the array arr[]
	Display(arr)

File: C-C++/Searching/LinearSearch/MoveToFrontHeadLinearSearch.c
// C program to implement the move to
// front to optimized Linear Search
#include <stdio.h>

// Structure of the array
struct Array {

	int A[10];
	int size;
	int length;
};

// Function to print the array element
void Display(struct Array arr)
{
	int i;

	// Traverse the array arr[]
	for (i = 0; i < arr.length; i++) {
		printf("%d ", arr.A[i]);
	}
	printf("\n");
}

// Function that swaps two elements
// at different addresses
void swap(int* x, int* y)
{
	// Store the value store at
	// x in a variable temp
	int temp = *x;

	// Swapping of value
	*x = *y;
	*y = temp;
}

// Function that performs the move to
// front operation for Linear Search
int LinearSearchMoveToFront(
	struct Array* arr, int key)
{
	int i;

	// Traverse the array
	for (i = 0; i < arr->length; i++) {

		// If key is found, then swap
		// the element with 0-th index
		if (key == arr->A[i]) {
			swap(&arr->A[i], &arr->A[0]);
			return i;
		}
	}
	return -1;
}

// Driver code
int main()
{
	// Given array arr[]
	struct Array arr
		= { { 2, 23, 14, 5, 6, 9, 8, 12 },
			10,
			8 };

	printf("Elements before Linear"
		" Search Move To Front: ");

	// Function Call for displaying
	// the array arr[]
	Display(arr);

	// Function Call for Move to
	// front operation
	LinearSearchMoveToFront(&arr, 14);

	printf("Elements after Linear"
		" Search Move To Front: ");

	// Function Call for displaying
	// the array arr[]
	Display(arr);
	return 0;
}
